### PR TITLE
Update core-obect (#5547, rebased)

### PR DESCRIPTION
**Pull request by nathanhammond**
Duplicating just in case. Don't want to lose things with a force push.

### CODE REVIEW

File: tests/unit/models/blueprint-test.js [hunk 1]
```diff
@@ -496,70 +559,158 @@ help in detail');
...
+        expect(output.shift()).to.match(/^installing/);
+        expect(output.shift()).to.match(/create.* .ember-cli/);
+        expect(output.shift()).to.match(/create.* .gitignore/);
+        expect(output.shift()).to.match(/create.* bar/);
+        expect(output.shift()).to.match(/create.* foo.txt/);
+        expect(output.shift()).to.match(/create.* test.txt/);
+        expect(output.length).to.equal(0);
+        var blueprintNew = new Blueprint(basicNewBlueprint);
+
+        ui.waitForPrompt().then(function() {
```
Comment (rwjblue) [inline tests/unit/models/blueprint-test.js]: This fails in CI for `MockUI` not including `waitForPrompt`.

File: tests/unit/models/blueprint-test.js [hunk 2]
```diff
@@ -574,6 +725,15 @@ help in detail');
 
           var blueprintNew = new Blueprint(basicNewBlueprint);
 
+          ui.waitForPrompt().then(function() {
```
Comment (rwjblue) [inline tests/unit/models/blueprint-test.js]: This fails in CI for `MockUI` not including `waitForPrompt`.

**Comment by rwjblue** [comment]
I commented on where the failure is coming from on those two failing tests.

**Comment by nathanhammond** [comment]
Conflict is here: https://github.com/ember-cli/ember-cli/commit/99227da548daba0768ced042d15a10819b0ac83e

**Comment by nathanhammond** [comment]
@stefanpenner can you review and r+?

File: blueprints/addon/index.js [hunk 3]
```diff
@@ -106,7 +106,7 @@ module.exports = {
   },
 
   mapFile: function(file, locals) {
-    var result = this._super.mapFile.call(this, file, locals);
+    var result = this._super(file, locals);
```
Comment (rwjblue) [inline blueprints/addon/index.js]: Should be updated to `this._super.mapFile.apply(this, arguments);` (we should only use bare `this._super()` in `init`).

File: lib/models/addon-discovery.js [hunk 4]
```diff
@@ -20,241 +20,238 @@ var findup     = require('findup');
...
-  @method discoverFromProjectItself
- */
-AddonDiscovery.prototype.discoverFromProjectItself = function(project) {
-  if (project.isEmberCLIAddon()) {
-    var addonPkg = this.discoverAtPath(project.root);
-    if (addonPkg) {
-      return [addonPkg];
+var AddonDiscovery = CoreObject.extend({
+  init: function(ui) {
+    this._super.init.apply(this, arguments);
```
Comment (rwjblue) [inline lib/models/addon-discovery.js]: Should be `this._super.apply(this, arguments);` (emulating ES6 classes where `super()` is used in constructor, but `super.someMethod()` is used everywhere else).

File: lib/models/addon.js [hunk 5]
```diff
@@ -65,772 +79,756 @@ function mergeTrees(inputTree, options) {
...
-    var message = name + ' is deprecated, please use ' +
-      insteadUse + ' directly instead  [addon: ' + addon.name + ']';
-
-    this.ui && this.ui.writeDeprecateLine(message);
-
-    if (!this.ui)  {
-      console.warn(message);
+var Addon = CoreObject.extend({
+  init: function(parent, project) {
+    this.parent = parent;
```
Comment (rwjblue) [inline lib/models/addon.js]: Should call `this._super.apply(this, arguments);` before touching `this`.

File: lib/models/addons-factory.js [hunk 6]
```diff
@@ -15,52 +15,50 @@ var debug      = require('debug')('ember-cli:addons-factory');
   @extends CoreObject
   @constructor
 */
-function AddonsFactory(addonParent, project) {
-  this.addonParent = addonParent;
-  this.project = project;
-}
+var AddonsFactory = CoreObject.extend({
+  init: function(addonParent, project) {
+    this.addonParent = addonParent;
```
Comment (rwjblue) [inline lib/models/addons-factory.js]: Should call `this._super.apply(this, arguments);` before touching `this`.

**Comment by rwjblue** [comment]
Commented on a few nits, but this is looking really good.

**Comment by nathanhammond** [comment]
@rwjblue I can make those changes. Why appears to be "to match ES6 patterns." (So that I'm not just cargo-culting.)

**Comment by rwjblue** [comment]
> Why appears to be "to match ES6 patterns."

Confirm.

**Comment by nathanhammond** [comment]
Rebased again because of #5739, updated to address @rwjblue's noticing of `_super` issues.

**Comment by nathanhammond** [comment]
@stefanpenner Can you provide commentary as to why https://github.com/ember-cli/ember-cli/pull/5737/commits/0d81b844dd3361db01f6e7de170100795344213d makes tests pass? Looks like some issue in core-object.

```
Cannot read property 'initializeAddons' of undefined
TypeError: Cannot read property 'initializeAddons' of undefined
    at CoreObject.Project.initializeAddons (/home/travis/build/ember-cli/ember-cli/lib/models/project.js:9:12709)
        at Project.initializeAddons (/home/travis/build/ember-cli/ember-cli/lib/models/project.js:9:12710)
        at Project.eachAddonCommand (/home/travis/build/ember-cli/ember-cli/lib/models/project.js:9:14665)
        at module.exports (/home/travis/build/ember-cli/ember-cli/lib/cli/lookup-command.js:9:1800)
        at CLI.<anonymous> (/home/travis/build/ember-cli/ember-cli/lib/cli/cli.js:9:2039)
        at lib$rsvp$$internal$$tryCatch (/home/travis/build/ember-cli/ember-cli/node_modules/rsvp/dist/rsvp.js:1036:16)
        at lib$rsvp$$internal$$invokeCallback (/home/travis/build/ember-cli/ember-cli/node_modules/rsvp/dist/rsvp.js:1048:17)
        at lib$rsvp$$internal$$publish (/home/travis/build/ember-cli/ember-cli/node_modules/rsvp/dist/rsvp.js:1019:11)
        at lib$rsvp$asap$$flush (/home/travis/build/ember-cli/ember-cli/node_modules/rsvp/dist/rsvp.js:1198:9)
        at nextTickCallbackWith0Args (node.js:420:9)
```

**Comment by stefanpenner** [comment]
its green!

**Comment by stefanpenner** [comment]
@homu r+ merge quickly!!

**Comment by nathanhammond** [comment]
@homu r-

**Comment by nathanhammond** [comment]
I believe there is an issue with core object.

1. [Test run with just rebase](https://travis-ci.org/ember-cli/ember-cli/builds/120634665)
2. [Applied this patch](https://github.com/ember-cli/ember-cli/pull/5737/commits/8e6176acde07a62850fd7933aee39064a9cc2db0)
3. [Test run with rebase and ES6 fixes](https://travis-ci.org/ember-cli/ember-cli/builds/120638616)

**Comment by nathanhammond** [comment]
I've cleaned up the commits and am in favor of landing this PR. I'll open an issue to investigate why we're unable to call `._super()` inside of `Addon#init` and `AddonsFactory#init`. That seems like a bug, but it is at least pre-existing based upon the diff from this commit.

I'm going to r+ assuming tests pass.

**Comment by nathanhammond** [comment]
Also we should discuss when we cut to beta (at tomorrow's meeting?). After landing this seems like a great time to do so.

File: lib/models/addon.js (hunk 5 above)
Comment (nathanhammond) [inline lib/models/addon.js]: Added https://github.com/ember-cli/ember-cli/issues/5750 to track this.

File: lib/models/addons-factory.js (hunk 6 above)
Comment (nathanhammond) [inline lib/models/addons-factory.js]: Added https://github.com/ember-cli/ember-cli/issues/5750 to track this.

**Comment by stefanpenner** [comment]
@homu r+

**Pull request merged**